streaming-payments.md

```diff
@@ -0,0 +1,375 @@
+# Streaming Tutorial
+
+## What you need before you start:
+
+* complete the [Letter Shop](./letter-shop) tutorial first
+
+## What you'll learn:
+
+* convert the proxy from the Letter Shop tutorial into a payment tool
+* using the `Pay` header in your shop and your payment tool
+* using the ILP packet
+* streaming payments
+* how to use trustlines to speed things up
+* Bilateral Transfer Protocol (BTP) and its relation to ILP
+
+## The Pay Header
+
+We'll change the Letter Shop from the previous tutorial a bit, to `shop2.js`:
+```js
+const IlpPacket = require('ilp-packet')
+const http = require('http')
+const crypto = require('crypto')
+const Plugin = require('ilp-plugin-xrp-escrow')
+function base64 (buf) { return buf.toString('base64').replace(/\+/g, '-').replace(/\//g, '_').replace(/=+$/, '') }
+function hash (secret) { return crypto.createHash('sha256').update(secret).digest() }
+function hmac (secret, input) { return crypto.createHmac('sha256', secret).update(input).digest() }
+
+let users = {}
+
+const plugin = new Plugin({
+  secret: 'ssGjGT4sz4rp2xahcDj87P71rTYXo',
+  account: 'rrhnXcox5bEmZfJCHzPxajUtwdt772zrCW',
+  server: 'wss://s.altnet.rippletest.net:51233',
+  prefix: 'test.crypto.xrp.'
+})
+
+plugin.connect().then(function () {
+  plugin.on('incoming_prepare', function (transfer) {
```

Comment:
Since you're not keeping track of incoming transfers, duplicate notifications (which can happen because the plugin interface doesn't stipulate that notifications must be delivered exactly once) would cause you to give out more letters than you intended.

Reply:
> the plugin interface doesn't stipulate that notifications must be delivered exactly once

Wait, is that the case? So then, even though the plugin already has to keep a full log of past transfer id's to be able to throw DuplicateIdError, the application code also needs to keep a list of non-expired transfer id's, to deduplicate `incoming_prepare` events? I never interpreted the LPI that way before, and I also think that's not how we have implemented apps that use plugins so far?

Reply:
I believe Evan is wrong - the current implementations are mostly at-most-once, not at-least-once for LPI events. This is because most plugins don't retry notifications at the moment.

That said, I don't think that is clear from the LPI docs right now and it should be. We should either say that plugins must provide at-most-once or exactly-once. OR we should say that plugins must provide either at-least-once or exactly-once. I'd prefer the former since that's the current behavior.